R/data.table.R

```diff
@@ -3086,3 +3110,150 @@ isReallyReal = function(x) {
   names(on) = xCols
   return(list(on = on, ops = idx_op))
 }
+
+top.n = function(n, wt, by, ties = F) {
+  ## helper that takes takes top.n() from the isub and massages it to produce indices for isub
+  
+  #' @param n must be non-zero integer.
+  #' @param wt optional unquoted column. 
+  #' @param by optional; accepts anything that the normal by argument accepts.
+  #' @param ties logical. Is ignored if wt is not provided.
+  #' @return List. The number of list elements depends on the arguments above. 
+  #' 
+  #' `top.n(n)` returns an unnamed list that must be evaluated in the parent frame for .N.
+  #' `top.n(n, wt)` returns a list to be called using the forder optimization
+  #' 
+  #' All other combinations return a named list to be evaluated as `do.call(`[data.table`, l_args)`
+
+  if (missing(n) || n == 0) {
+    stop("A non-zero n argument needs supplied to top.n(n, wt, by, ties = F)")
+  }
+
+  l_args = list(x = quote(x))
+  if (!missing(by))  l_args[['by']] = substitute(by)
+  
+  if (missing(wt)){ 
+    if (ties) warning("Ties argument is TRUE without a wt column supplied. Ties argument will be ignored - head(dt, n) or, if n < 0, tail(dt, abs(n)) will be returned")
+    
+    if (n < 0) ind_sub = substitute(max(1, .N + n + 1):.N, list(n = n)) else ind_sub = substitute(1:min(n, .N), list(n = n)) 
+    
+    if (missing(by)){
+      return(list(substitute(ind <- inds, list(inds = ind_sub)))) 
+    } else {
+      l_args[['j']] = substitute(.I[inds], list(inds = ind_sub))
+    }
+  
+  } else {
+    
+    if (ties){
+      if (n < 0){
+        l_args[['j']] = substitute(.I[frank(wt, ties.method = 'min') <= n], list(n = abs(n), wt = substitute(wt)))
+      } else {
+        l_args[['j']] = substitute(.I[frank(-wt, ties.method = 'min') <= n], list(n = n, wt = substitute(wt)))
+      } 
+    } else {
+
+      i_j_subset = substitute(1:min(n, .N), list(n = abs(n)))
+      if (n < 0) i_order = substitute(order(wt), list(wt = substitute(wt))) else i_order = substitute(order(-wt), list(wt = substitute(wt)))
+      
+      if (missing(by)){
+        l_args[['i']] = substitute(order_i[i_subset], list(order_i = i_order, i_subset = i_j_subset))
+        l_args[['f_order']] = TRUE
+      } else {
+        l_args[['i']] = i_order
+        l_args[['j']] = substitute(.I[j_subset], list(j_subset = i_j_subset))
+      }
+    }
+  }
+
+  l_args
+}
+
+filter.at = function(cols, logic, by, all.vars = TRUE){
+  ## helper that takes takes filter.at() from the isub and massages it to produce indices for isub
+  
+  #' @param cols required; can accept all values that .SDcols accepts
+  #' @param logic required; a function or unquoted text that results in logic evaluation. The
+  #'              unquoted text must include `x` as an argument.
```

Comment:
> unquoted text must include `x` as an argument.

why you chose such an API? I haven't seen it anywhere in base R and package that I am using. IMO it is better to at least expect `function(x) ...logic...` then `x` is a naturally expected object in the _logic_. User may have `x` variable defined in their code and expect it to be found in current scenario. This is how scoping in R should behave, proposed API works against it.

Reply:
```purrr``` uses ```~ .x``` shorthand - I admit that I may have have taken it to the extreme.

Reply:
It is counter inituitive, the purr approach, because `~` has quite different context in base R. And yes, your approach is even more...